aten/src/ATen/native/mps/operations/Copy.mm

```diff
@@ -0,0 +1,220 @@
+//  Copyright © 2022 Apple Inc.
+
+#include <ATen/mps/MPSStream.h>
+#include <ATen/native/mps/Copy.h>
+#include <ATen/native/mps/OperationUtils.h>
+#include <iostream>
+#include <cstring>
+#include <ATen/ATen.h>
+#include <ATen/Tensor.h>
+#include <ATen/Utils.h>
+#include <torch/library.h>
+#include <ATen/native/Resize.h>
+#include <c10/util/Optional.h>
+
+namespace at {
+namespace native {
+namespace mps {
+
+void* pageAlignedBlockPtr(
+    const void* ptr,
+    NSUInteger size,
+    NSUInteger* alignedBlockSize) {
+  uintptr_t address = (uintptr_t)ptr;
+  uintptr_t alignedAddress = address & ~(PAGE_SIZE - 1);
+  uintptr_t alignedEnd = ((address + size) + PAGE_SIZE - 1) & ~(PAGE_SIZE - 1);
+  uint64_t alignedLength = alignedEnd - alignedAddress;
+
+  assert(address >= alignedAddress);
+  assert(address + size <= alignedAddress + alignedLength);
+
+  *alignedBlockSize = alignedLength;
+  return (void*)alignedAddress;
+}
+
+static at::Tensor& copy_from_mps_(at::Tensor& self, const at::Tensor& src,
+                           bool non_blocking) {
+
+  id<MTLDevice> device = MPSDevice::getInstance()->device();
+  MPSStream* stream = getCurrentMPSStream();
+  uint64_t size = src.nbytes();
+  if (size == 0) return self;
+  void* host_dst = self.data_ptr();
+
+  // MTLContext* context = static_cast<MTLContext *>(device->device_handle);
+  auto storage_byte_offset = src.storage_offset() * src.itemsize();
+  id<MTLBuffer> sourceBuffer = __builtin_bit_cast(id<MTLBuffer>, src.storage().data());
+  if (sourceBuffer == nil) return self;
+  NSUInteger destOffset = 0;
+
+  @autoreleasepool {
+    MTLResourceOptions options = MTLResourceOptionCPUCacheModeDefault | MTLResourceStorageModeShared;
+    NSUInteger alignedLength = 0;
+
+    void* alignedPtr = pageAlignedBlockPtr(host_dst, (NSUInteger)size, &alignedLength);
+    id<MTLBuffer> destBuffer = [device newBufferWithBytesNoCopy:alignedPtr
+                                                         length:alignedLength
+                                                        options:options
+                                                    deallocator:nil];
+    destOffset = uintptr_t(host_dst) - uintptr_t(alignedPtr);
+    // 4 bytes alignment required on macos for blits.
+    TORCH_CHECK(destOffset % 4 == 0, "Unaligned blit request");
+
+    dispatch_sync(stream->queue(), ^() {
+      @autoreleasepool {
+        id<MTLCommandBuffer> commandBuffer = stream->commandBuffer();
+        id<MTLBlitCommandEncoder> blitEncoder =
+            [commandBuffer blitCommandEncoder];
+
+        [blitEncoder copyFromBuffer:sourceBuffer
+                       sourceOffset:(NSUInteger)storage_byte_offset
+                           toBuffer:destBuffer
+                  destinationOffset:(NSUInteger)destOffset
+                               size:(NSUInteger)size];
```

Comment:
contiguous only, it would seem

Reply:
Yes, in this PR I didn't include the Scatter/Gather kernel to handle the view operation as it was hard to split it with other operations support. But I will update it here in a commit.

Reply:
Yeah, so this is one of those situations where I would have preferred to just have the entire PR, instead of having it split up this way, because then I can crossref what the plan is with the actual use sites.

Reply:
@ezyang , I have uploaded as_strided op in the PR itself. You can take a look, we use the MPSGraphCache there and has the Gather operation as part of it.